src/core/border_router/routing_manager.cpp

```diff
@@ -730,62 +844,214 @@ void RoutingManager::HandleRouterAdvertisement(const Ip6::Address &aSrcAddress,
     while ((option = Option::GetNextOption(option, optionsBegin, optionsLength)) != nullptr)
     {
         const PrefixInfoOption *pio;
-        Ip6::Prefix             prefix;
+        const RouteInfoOption * rio;
 
-        if (option->GetType() != Option::Type::kPrefixInfo)
+        switch (option->GetType())
         {
-            continue;
+        case Option::Type::kPrefixInfo:
+            pio = static_cast<const PrefixInfoOption *>(option);
+            if (pio->IsValid())
+            {
+                needReevaluate |= UpdateDiscoveredPrefixes(*pio);
+            }
+            break;
+
+        case Option::Type::kRouteInfo:
+            rio = static_cast<const RouteInfoOption *>(option);
+            if (rio->IsValid())
+            {
+                needReevaluate |= UpdateDiscoveredPrefixes(*rio);
+            }
+            break;
+
+        default:
+            break;
         }
+    }
+
+    if (needReevaluate)
+    {
+        EvaluateRoutingPolicy();
+    }
+
+exit:
+    return;
+}
+
+bool RoutingManager::UpdateDiscoveredPrefixes(const RouterAdv::PrefixInfoOption &aPio)
+{
+    Ip6::Prefix prefix;
+    bool        needReevaluate = false;
+
+    aPio.GetPrefix(prefix);
+    if (!IsValidOnLinkPrefix(prefix))
+    {
+        otLogInfoBr("ignore invalid prefix in PIO: %s", prefix.ToString().AsCString());
+        ExitNow();
+    }
+
+    otLogInfoBr("discovered on-link prefix (%s, %u seconds) from interface %u", prefix.ToString().AsCString(),
+                aPio.GetValidLifetime(), mInfraIfIndex);
+
+    if (aPio.GetValidLifetime() == 0)
+    {
+        needReevaluate |= (InvalidateDiscoveredPrefixes(&prefix, /* aIsOnLinkPrefix */ true) > 0);
+    }
+    else
+    {
+        needReevaluate |= AddDiscoveredPrefix(prefix, /* aIsOnLinkPrefix */ true, aPio.GetValidLifetime());
+    }
+
+exit:
+    return needReevaluate;
+}
+
+bool RoutingManager::UpdateDiscoveredPrefixes(const RouterAdv::RouteInfoOption &aRio)
+{
+    Ip6::Prefix prefix;
+    bool        needReevaluate = false;
+
+    aRio.GetPrefix(prefix);
+    if (!IsValidOmrPrefix(prefix))
```

Comment:
Should we auto-install the external route for any RA(RIO) learned from infrastructure link? like a prefix that might be of 48 bit length

Reply:
add external route for RIO is for connectivity to another Thread network which is connected to the same infra link. I don't feel it necessary to add explicit support for non-duckhorn BR prefixes in another Thread network. It there is no duckhorn BR in that Thread network, it would be an expected result that there is no connectivity to the Thread network. Thoughts?

Reply:
Fine to me to keep it as is.